### DISCUSSION TITLE
How to custom sort with DataGrid?

### DISCUSSION THREAD

**Discussion by JerryJian** [AvaloniaUI/Avalonia#14132] [Q&A]
How to custom sort with DataGrid? 

When I want to do something in the  `Sorting` event of `DataGrid`, 
1. I can't get the current state is Ascending or Descending.
2. If I set the e.Handled to true, the sort indicator will not be changed, the pseudo classes is not updated.
3. If I update the pseudo classes by code, the column header will recaculate it when PointerEnter/PointerExited.

The HeaderCell of e.Column is internal.

So how can I custom the sort logic of DataGrid?

**Comment by workgroupengineering** [comment]
1. For custom sorting logic you implement in your column type  [IComparable\<T> Interface](https://learn.microsoft.com/en-us/dotnet/api/system.icomparable-1?view=net-8.0).
2. To get the sorting status, monitor your DataGridCollectionView

https://github.com/AvaloniaUI/Avalonia/blob/38e839997d2c204548e6fad396c178780a010cb1/src/Avalonia.Controls.DataGrid/Collections/DataGridCollectionView.cs#L939

**Reply by JerryJian** [reply]
I have a collection of 1000 entries, displaying 50 entries per page using DataGrid. Is there any way to sort the entire set when I click on column sorting? Thank you

**Reply by timunie** [reply]
Normally it's no issue to have 1000 items at once in the DG as the rows are virtualized.

**Reply by asmirnov82** [reply]
I would like to impelement server side sorting and paging. So I would like to intercept Sorting event from the column and write my own login of fetching new data from the server to display different page. In this case impementing IComparable<T> wouldn't help as it will only sort loaded page
